core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -993,28 +1019,40 @@ public void updateBufferSize(ByteSizeValue shardIndexingBufferSize, ByteSizeValu
         if (preValue.bytes() != shardIndexingBufferSize.bytes()) {
             // so we push changes these changes down to IndexWriter:
             engine.onSettingsChanged();
+            logger.debug("updating index_buffer_size from [{}] to [{}]", preValue, shardIndexingBufferSize);
+
+            long iwBytesUsed = engine.indexWriterRAMBytesUsed();
+            if (shardIndexingBufferSize.bytes() < iwBytesUsed) {
+                // our allowed buffer was changed to less than we are currently using; we ask IW to refresh
+                // so it clears its buffers (otherwise it won't clear until the next indexing/delete op)
+                logger.debug("refresh because index buffer decreased to [{}] and IndexWriter is now using [{}] bytes",
+                             shardIndexingBufferSize, iwBytesUsed);
 
-            if (shardIndexingBufferSize == EngineConfig.INACTIVE_SHARD_INDEXING_BUFFER) {
-                // it's inactive: make sure we do a refresh / full IW flush in this case, since the memory
-                // changes only after a "data" change has happened to the writer
-                // the index writer lazily allocates memory and a refresh will clean it all up.
-                logger.debug("updating index_buffer_size from [{}] to (inactive) [{}]", preValue, shardIndexingBufferSize);
+                // TODO: should IW have an API to move segments to disk, but not refresh?
```

Comment:
I'm curious about this one - I thought the segments are always (once made) on disk - but not fsynced?